### ISSUE TITLE
i1 cases are not handled in Linalg on tensors pipeline

### ISSUE THREAD

**Issue by hanhanW** [google/iree#5153]
**Describe the bug**
I actually don't expect to see failures in this directory when setting the flags on...

I don't collect all the failures because the build system will stop when there is an error. I suspect the issue is how we deal with `i1` cases. 

In Linalg on buffers path, we will replace i1 with i8 in MaterializeInterfacesPass.

In Linalg on tensors path, we don't do it because DispatchLinalgOnTensorsPass (which is run before MaterializeInterfacesPass) will generate something like `flow.dispatch.tensor<readonly:4xi1>` (which is one of the inputs of the op).

i1 is always an annoying issue...

**To Reproduce**

Run `iree-translate --iree-mlir-to-vm-bytecode-module --iree-hal-target-backends=vulkan-spirv iree/test/e2e/tosa_ops/select.mlir -o a.vmfb -iree-flow-dispatch-linalg-on-tensors=true -iree-codegen-spirv-experimental-linalg-on-tensors=true`

cc @rsuderman @NatashaKnk

**Comment by hanhanW** [comment]
Interesting, we hit the same issue in some integration tests

```
module  {
  func @greater__2x2x2_2x2x2__f32__uniform(%arg0: tensor<2x2x2xf32>, %arg1: tensor<2x2x2xf32>) -> tensor<2x2x2xi1> attributes {iree.module.export} {
    %0 = "mhlo.compare"(%arg0, %arg1) {comparison_direction = "GT"} : (tensor<2x2x2xf32>, tensor<2x2x2xf32>) -> tensor<2x2x2xi1>
    return %0 : tensor<2x2x2xi1>
  }
  func @greater__2x2x2_2x2x2__i32__uniform(%arg0: tensor<2x2x2xi32>, %arg1: tensor<2x2x2xi32>) -> tensor<2x2x2xi1> attributes {iree.module.export} {
    %0 = "mhlo.compare"(%arg0, %arg1) {comparison_direction = "GT"} : (tensor<2x2x2xi32>, tensor<2x2x2xi32>) -> tensor<2x2x2xi1>
    return %0 : tensor<2x2x2xi1>
  }
}
```

**Comment by hanhanW** [comment]
I remember the conclusion was IREE don't want to see any i1 types, they should be converted to i8. I'm not sure if this is still the case.

**Comment by benvanik** [comment]
I was hitting some things with this the other day too. We need to fix i1 before dispatch region formation: really, we should be widening i1 to i8 (or higher) very early on (maybe in linalg, maybe before).

There are two major i1 cases: scalar i1 tensors (like this) because we don't yet have #1159 to turn them into i1 primitives like they should be, and non-scalar i1 tensors (masks, etc) that have actual wide math performed on them.

The current way of just casting larger non-scalar tensors is really wasteful (7x waste) as it's just widening the type, when really we should be treating Nxi1 as a bitvector. This would be a really good optimization to have in linalg: a loop over Nxi1 could be changed to a loop over N/8 i8, which then could be vectorized further down, etc.

The workaround until we have detensoring (and maybe something still required after it) is to insert the appropriate casts where required like the old path is. We could also turn on the implicit conversion that was turned off for linalg-on-tensors here:
https://github.com/google/iree/blob/b327096ae2df93b3441d63ed3fbe5ee8a51d0fe9/iree/compiler/Dialect/Flow/Transforms/LegalizeInputTypes.cpp#L88-L93
I forget why - may have just been ntv wanting to complete the prototype - but that's probably the first place I'd look until we solve detensoring and vectorization of i1s.

**Comment by MaheshRavishankar** [comment]
To have Linalg vectorize this correctly, etc. we will have to do this after this https://github.com/google/iree/issues/5045 . I think its WIP and pretty close to landing. This will then enable using vectorization for the element-wise operations as well, and by extension for i1 ops.

**Comment by hanhanW** [comment]
I will take this one.

**Comment by hanhanW** [comment]
Thanks for the detailed context, btw!

**Comment by hanhanW** [comment]
I'm trying to make it work, but I hit issues.

The input example is

```mlir
func @narrow_int_i32_i1() -> tensor<3xi1> attributes { iree.module.export } {
  %input = iree.unfoldable_constant dense<[-42, 0, 42]> : tensor<3xi32>
  %res = "mhlo.convert"(%input) : (tensor<3xi32>) -> tensor<3xi1>
  return %res : tensor<3xi1>
}
```

The log of Linalg on tensors: https://gist.github.com/hanhanW/ec3c9ee8bc1706abb3eed47338b26a68
The log of Linalg on buffers: https://gist.github.com/hanhanW/4b19df69c76c10d7b389744654f2ef2e

Linalg on buffers works because it calls MaterializeInterfacesPass before converting ops to Linalg ops. We can not do similar thing to Linalg on tensors because flow.dispatch/flow.dispatch.workgroups stuff are not created before DispatchLinalgOnTensorsPass. Actually flow.dispatch.workgroups are created in DispatchLinalgOnTensorsPass, though.

To repro:

```
$ iree-translate --iree-mlir-to-vm-bytecode-module --iree-hal-target-backends=vulkan-spirv -iree-flow-dispatch-linalg-on-tensors=true -iree-codegen-spirv-experimental-linalg-on-tensors=true a.mlir -o a.vmfb
```

There are two approaches to me. One is to widen all i1s to i8s in Linalg ops, another is to apply vectorization of i1s in Linalg.

## Widen i1s to i8s in Linalg ops

The IR before DispatchLinalgOnTensorsPass is

```mlir
func @narrow_int_i32_i1() -> tensor<3xi1> attributes {iree.module.export, iree.reflection = {f = "I1!R8!B5!t4d3", fv = "1"}} {
  %cst = constant dense<[-42, 0, 42]> : tensor<3xi32>
  %c0_i32 = constant 0 : i32
  %0 = iree.do_not_optimize(%cst) : tensor<3xi32>
  %1 = linalg.init_tensor [3] : tensor<3xi1>
  %2 = linalg.generic {indexing_maps = [affine_map<(d0) -> (d0)>, affine_map<(d0) -> (d0)>], iterator_types = ["parallel"]} ins(%0 : tensor<3xi32>) outs(%1 : tensor<3xi1>) {
  ^bb0(%arg0: i32, %arg1: i1):  // no predecessors
    %3 = cmpi ne, %arg0, %c0_i32 : i32
    linalg.yield %3 : i1
  } -> tensor<3xi1>
  return %2 : tensor<3xi1>
}
```

I tried to rewrite the i1s to i8s, but I found that I will have to deal with operations in the region. And this is very painful because there will be special logics, like insert cast ops if it yields i1 types.

I'm still looking for a good place to insert cast ops.

## Vectorization of i1s

I don't have clear picture on it now. E.g., if we pack `Nxi1` into `(N/8)xi8`, what would it look like in Linalg? Is the type still `Nxi1` or the type will be `(N/8)xi8`. The former makes more sense to me because the latter one will have to handle indexing issues. The operand shapes change, so the indexing_maps also change. Otherwise it will be an invalid Linalg op because one AffineDimExpr will match to two different shape size.

If it is still i1, the issue will happen in downstream, like converting it to SPIR-V. This will be tricky because the type converter in SPIR-V does not recognize something like memref<i1x3>. We will have to figure out to support i1 types, and load/store i1 types in a proper way.

## Conclusion

I'm still exploring the possibilities and I will try to find a better place to insert cast ops tomorrow. I would be appreciate if there are some inputs for above investigation or other direction.

**Comment by benvanik** [comment]
Thanks for looking into this! It'll be really nice to be able to consistently apply things so that we can stop having to worry about i1s and the way you're approaching it will help us get there :)

Splitting the problem into "how do we represent i1 tensors in linalg ops" from "how do we handle load/store of i1 tensors outside of linalg ops" may make this easier. Framing this entirely in standard upstream dialects and nothing IREE-specific that could apply uniformly across the entire program makes it appealing - less dependencies on pass order, less assumptions about conversions during lowering ("this will get turned into i8 during flow->hal"), etc. It'd also let us remove the second question entirely: we _don't_ represent i1 tensors outside of dispatch regions, and could even verify that no i1 tensors exist after dispatch regions are formed.

The vectorization would gives us the smallest memory footprint, the least complexity in lowering to various backends, pave the way for other small non-byte-aligned types (i4), and improve performance of anything using dense i1s for masking (less memory bandwidth, SIMD, etc). Just widening i1->i8 like we do in the old path would unblock things but we may end up finding all the bugs we've squashed/worked around before again and it may take just as much time to get right :)

Assuming we can solve the way to take generic linalg IR with `tensor<...i1>` to generic linalg IR with `tensor<...iN>` then the remaining ops that deal with i1's outside of the dispatch regions would be things like `tensor.extract` to get an i1 out of a tensor, and that becomes the same kind of stuff as in LLVM's BitVector.h:
```c
    Bits[Idx / BITWORD_SIZE] |= BitWord(1) << (Idx % BITWORD_SIZE);
```
and
```c
    BitWord Mask = BitWord(1) << (Idx % BITWORD_SIZE);
    return (Bits[Idx / BITWORD_SIZE] & Mask) != 0;
```

which comes out to:

```mlir
%4 = tensor.extract %3[%idx] : tensor<32xi1>
```

which would become something like:

```mlir
// almost all of this folds away into other ops/constants
%idx_div_32 = divi %idx, %c32 : index
%idx_mod_32 = modi %idx, %c32 : index
%4 = tensor.extract %3[%idx_div_32] : tensor<1xi32>
%mask = shl %c1, %idx_mod_32 : i32
%4bit = andi %4, %mask : i32
%4 = cmpi ne, %4bit, %c0 : i1
```

Rewriting these standard ops doing tensor read/writes seems like a nicely separable task once the bulk of operations acting on `tensor<i1>` are turned into linalg ops - there should only be a few left.

In theory linalg (and the information it retains) should make the vectorization easy but it may be missing some helpers and std is probably missing some casts like you're finding. Maybe @nicolasvasilache has some ideas of utilities that may be useful/reusable.

**Comment by MaheshRavishankar** [comment]
This might conflict with Ben's comment (I will respond to that one later), but @hanhanW I dont see a way forward without having to do vectorization of i1 ops. 

1) How do we lower tensor<...xi1> . Maybe there is a way to do this in standard dialect itself, but my first thought that we do this in the SPIR-V dialect. Essentially all `memref<...xi1>` turn into `!spv.array<..xi8>` (or `!spv.array<...xi32>` depending on the capabilities). Then a `load %a[...] : tensor<...xi1>` becomes a 

```
%p = spv.AccessChain %a[..] : !spv.array<..xi8>
%v = spv.load %p
```

So you would load `i8` (or `i32`) values which themselves contain `8xi1` values (or `32xi1` values), and do some bit arithmatic to extract the right bit. This gets a functional way of loading handling `i1` tensors.

2) Next we can then try to vectorize these operations. We could use a vector length of 8 (or 32). Thomas already has all ops going through the tile + distribute + vectorize flow. So this should be a matter of adding a configuration for picking the vector length correctly

You will end up with `vector.transfer_read ... : memref<...xi1> to vector<8xi1>`. We might need to add an extra pattern to lower these to just a load of the `i8` when lowering to SPIR-V so that everything matches up type-wise.

3) The last step is how this is represented in IREE. During dispatch region creation, if any of the operands have an `i1` type, then we have to change the type to `i8` during the dispatch region creation.  I dont have the concrete steps here, but essentially the fact that this is in `i1` tensor should only be visible to dispatch itself, while outside the dispatch flow sees this as an `i8` tensor. This shouldnt be too hard. Within the dispatch region the `flow.dispatch.load` and `flow.dispatch.store` (the names might be different now) convert from Flow input/output type to tensor type. We can use this to hide this information and make it appear that inputs and outputs are `i8` to Flow, while maintaining `i1` inside the dispatch region.

I think to get things working we need (1) and (3). (2) becomes an optimization.

Let me know if you want to discuss this, and happy to pick up parts of this if it gets tricky.

**Comment by MaheshRavishankar** [comment]
> Assuming we can solve the way to take generic linalg IR with `tensor<...i1>` to generic linalg IR with `tensor<...iN>`

This would actually be fairly tricky, the indexing maps that you need to use might become invalid and it doesnt compose as well. We need to do "vectorize" this like any other operation vectorization. See above (https://github.com/google/iree/issues/5153#issuecomment-810556596)

**Comment by benvanik** [comment]
The issue is that we have things like `tensor.extract` (which becomes a `flow.tensor.load`) outside of the dispatch regions - whatever we do inside the regions thus has to be mirrored outside as well. If we instead did the vectorization before dispatch region formation we don't have anything special. This is strictly a phase ordering thing: if we do it after we partition the code both sides need to agree (and agree consistently across all backends), whereas if we do it before then nothing needs to know about it except the single transformation doing the conversion.

We should all chat more :)

**Comment by hanhanW** [comment]
Thanks for all the words! I think I mixed the context of inside/outside of dispatches. I agree that codegen should have ability to handle i1 cases, and we should also work on "how do we handle load/store of i1 tensors" at flow level.

So far we don't see these issues on CPU side because there are some bitcast ops in LLVM side, IIRC. And I guess i1s are materialized in ConvertToLLVMPass -- it queries things from `HALDispatchABI::getInputTypes`, so it allocates the correct size buffer. I was worried that the buffer sizes are not match on host side and device side, but it looks like we're handling well. :p [full log](https://gist.github.com/hanhanW/e1b5ebc0c79ff0b50978614a3a535689)

I did not address issues by adding support on SPIR-V side, because it was push back before: https://reviews.llvm.org/D81317
But now I think we can give it another shot. I will go with @MaheshRavishankar first point, because it looks like codegen should have ability to handle i1 cases. And I will revisit if there are any other needs.

For the third point from Mahesh, I now feel it is not needed, but we'll see. The variable creation logic is in ConvertToSPIRVPass.cpp. The pass itself will query from the type converter and allocate a buffer with proper size. (Actually, I'm not sure if "allocate a buffer" are right terms, but basically it will create spirv::GlobalVariableOp.)

> We can use this to hide this information and make it appear that inputs and outputs are i8 to Flow, while maintaining i1 inside the dispatch region.

It is still a good discussion to me, because it is more clear that where we should put the boundary. It looks like we put the logic in the type converter of the target, ie the last stage of codegen (_ConvertToXXXPass_). I'm probably able to "fix" it, we'll see. The questions now would be -- Do we miss anything at flow level? Will we hit the same issue again? If this is the way to go, we should document it somewhere.

I will start working on it, and will put more context here if I find something new!

**Comment by benvanik** [comment]
> Do we miss anything at flow level? Will we hit the same issue again?

having trouble following what the state is here, but if you're talking about doing the naive i1 -> i8 widening at the flow level things should work - and "working" may be all we want here - but if we want it to be good/fast/optimal then we want to pack things in linalg **before** dispatch region formation. If we want to support 4bit quantization - which we do - we'll need the packing anyway and may as well use the same approach for both. Think of the problem not as something specific to i1, but a problem of sub-byte-aligned data (i1, i2, i4, i23, etc). The same reason one does not store quantized 8 bit data in 32 bit ints is the same as to why we don't want to store 4 bit data in 8 bit ints or 1 bit data in 8 bit ints - it defeats the purpose of having the type not just be larger to begin with :)

It's of course good to hack in something that unblocks this path - but it'd be nice to have the larger solution understood a bit more so that we know what's a hack and what's incremental work towards something better and more robust.

**Comment by MaheshRavishankar** [comment]
I think there is agreement that we should pack the `tensor<...xi1>` into `tensor<...(N/8)xvector<8xi1>` (I am using the packed type for representation, not saying you need to use that type specifically). Ben and I were just discussing we need such packing/unpacking ops. If you think you can support these, I would be totally on board with this. My suggestion is 
1) Try to do the packing if it is something that doesnt hit very esoteric problems
2) If you end up going down many rabbit holes, cut your losses and have a working solution first (so that we can unblock linalg on tensors) and then go back to doing it with packing.

**Comment by antiagainst** [comment]
> I have a workaround in SPIR-V, which unconditionally treats all i1s are i8s (https://reviews.llvm.org/D81317). But I think we would like to have real packing, like `11100000`, not `10000000 10000000 10000000`. (@antiagainst please correct me if I'm wrong.)

Yup. Bool is an abstract type in SPIR-V and it does not have defined bitwidth. Having it in interface storage classes (across ABI) is generally painful. Better to avoid if possible; but otherwise, packing is certainly preferable. SGTM if you want to do it in SPIR-V. I recall you actually sent [patches](https://reviews.llvm.org/D78974) before for this kind of bit width packing/emulation and the logic there can be reused. (But again SPIR-V is just half of the world. Right now we perform such packing/enumeration unconditionally, which might run in conflict with what the runtime side want. We need to have an option to the type converter to control it. Basically [this TODO](https://github.com/llvm/llvm-project/blob/dc4e913be9c3d1c37f66348d4b5047a107499b53/mlir/lib/Dialect/SPIRV/Transforms/SPIRVConversion.cpp#L249-L255). You don't need to do it though; I can handle it later.)

**Comment by hanhanW** [comment]
Thanks! I will go with storing an i1 to in a 8 bits buffer in SPIR-V, so we can unblock the issue.

I have a bigger picture now. I was struggled in data layout (like how would we actually store i1s for now and future). I think this is similar to `std::vector<bool>`, so we can make it space efficient. Today, IREE and llvm-aot stores an i1 into an i8 buffer. With some hacks, I can make it happen, but there are some mysteries in llvm-aot path to me. llvm-aot always widen i1 to i8.

I think all tensor<...xi1> are happened on device side. The host side would not have them unless there are i1 constants. But we will hit the issue because the check tests wanna check the i1 values.

To make it happy, my hack is:

1. Revert #3066
2. Add support for `serializeConstantI1Array` in iree/compiler/Dialect/VM/Target/Bytecode/ConstantEncoder.cpp

https://github.com/google/iree/blob/b7381623652bdcfd67ad2cf836f503c9d634c567/iree/compiler/Dialect/VM/Target/Bytecode/ConstantEncoder.cpp#L26-L47

I think it is a good option to let type converter to control "how to handle with i1s". They can be real one bit per i1, or 8 bits per i1.

For IREE, we probably should have a mechanism to configure backends "how to store i1s". So the assumptions in host side and device side will not mismatch.

**Comment by hanhanW** [comment]
I just found that the hack I had for storing one bit per i1 doesn't work anymore because of #5239... It is not easy to me to make IREE happy again.

I think we should teach SPIR-V to do what today does, ie storing i1 to i8 buffer. Migrate to Linalg on tensors. And applied what the packing mechanism we discussed yesterday. (Store one i1 in one bit).

For the last step, we should make sure

1. CPU backend can handle packing problem
2. GPU backend can handle packing problem
3. IREE hal can store i1 constants correctly.
4. Teach python binding to load i1s or widen i1 to i8 before going to binding's world.

**Comment by MaheshRavishankar** [comment]
@hanhanW for the immediate needs (to get passed this on the linalg on tensors path and not be blocked by this), everything is in place to handle this right?

**Comment by hanhanW** [comment]
yes!

**Comment by MaheshRavishankar** [comment]
do we have enough lit tests for these. I have a feeling we have always been scared of adding lit tests for cases cause the fixes would be really hard. If we are confident now about this, then worth trying.

**Comment by GMNGeoffrey** [comment]
I hit another issue: https://github.com/google/iree/issues/3102#issuecomment-963409877

**Comment by MaheshRavishankar** [comment]
As a fallback `i1` tensors are expanded to `i8` and handled in the backend. This issue can be closed now.
